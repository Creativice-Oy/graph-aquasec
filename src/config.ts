import {
  IntegrationExecutionContext,
  IntegrationInstanceConfigFieldMap,
  IntegrationInstanceConfig,
} from '@jupiterone/integration-sdk-core';

/**
 * A type describing the configuration fields required to execute the
 * integration for a specific account in the data provider.
 *
 * When executing the integration in a development environment, these values may
 * be provided in a `.env` file with environment variables. For example:
 *
 * - `CLIENT_ID=123` becomes `instance.config.clientId = '123'`
 * - `CLIENT_SECRET=abc` becomes `instance.config.clientSecret = 'abc'`
 *
 * Environment variables are NOT used when the integration is executing in a
 * managed environment. For example, in JupiterOne, users configure
 * `instance.config` in a UI.
 */
export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  /**
   * NOTE: Aqua calls this an "API Key", but it is essentially an auto-generated
   * API key _name_. The value of this property is _visible_ from the Aqua
   * console. The actual sensitive token is the `apiSecret` described below.
   */
  apiKey: {
    type: 'string',
  },
  apiSecret: {
    type: 'string',
    mask: true,
  },
  accountId: {
    type: 'string',
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  apiKey: string;
  apiSecret: string;
  accountId: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationInstanceConfig>,
) {
  await Promise.resolve();
}
