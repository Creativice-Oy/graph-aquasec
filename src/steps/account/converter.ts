import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { AquaSecAccount } from '../../aquasec/types';

import { Entities } from '../constants';

export function createAccountEntity(data: AquaSecAccount): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: `aquasec_account:${data.id}`,
        _type: Entities.ACCOUNT._type,
        _class: Entities.ACCOUNT._class,
        id: data.id.toString(),
        name: data.id.toString(),
        owner: data.root_user,
        createdOn: parseTimePropertyValue(data.created),

        stripeId: data.stripe_id,
        trialEnd: parseTimePropertyValue(data.trial_end),
        currentPlan: data.current_plan,
        additionalCredits: data.additional_credits,
        numAllowedKeys: data.num_allowed_keys,
        suppressionsAdminOnly: data.suppressions_admin_only,
        globalSuppressionsAdminOnly: data.global_suppressions_admin_only,
        source: data.source,
        enforceOauth: data.enforce_oauth,
        usersAdminOnly: data.users_admin_only,
        integrationsAdminOnly: data.integrations_admin_only,
        disableDefaultEmails: data.disable_default_emails,
        blockDefaultGroup: data.block_default_group,
        presuppressPlugins: data.presuppress_plugins,
        associateTags: data.associate_tags,
        awsMarketplaceId: data.aws_marketplace_id,
        cspEnabled: data.csp_enabled,
        rootUser: data.root_user,
      },
    },
  });
}
