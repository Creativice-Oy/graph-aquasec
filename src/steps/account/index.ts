import {
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';
import { createApiClient } from '../../aquasec/jupiterone';
import { IntegrationConfig } from '../../config';
import { setAccountEntity } from '../../util/jobState';
import { Steps, Entities } from '../constants';
import { createAccountEntity } from './converter';

export async function fetchAccountDetails(
  context: IntegrationStepExecutionContext<IntegrationConfig>,
) {
  const { jobState, instance } = context;
  const client = createApiClient(context);
  const account = await client.getAccount(instance.config.accountId);

  await setAccountEntity(
    jobState,
    await jobState.addEntity(createAccountEntity(account)),
  );
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: Steps.ACCOUNT,
    name: 'Fetch Account Details',
    entities: [Entities.ACCOUNT],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
