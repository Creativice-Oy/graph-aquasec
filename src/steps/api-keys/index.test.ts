import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../../config';
import { integrationConfig, withRecording } from '../../../test/config';
import { fetchAccountDetails } from '../account';
import { fetchApiKeys } from '.';

describe('#fetchApiKeys', () => {
  test(
    'should collect data',
    withRecording('fetchApiKeys', __dirname, async () => {
      const context = createMockStepExecutionContext<IntegrationConfig>({
        instanceConfig: integrationConfig,
      });

      await fetchAccountDetails(context);
      await fetchApiKeys(context);

      expect({
        numCollectedEntities: context.jobState.collectedEntities.length,
        numCollectedRelationships:
          context.jobState.collectedRelationships.length,
        collectedEntities: context.jobState.collectedEntities,
        collectedRelationships: context.jobState.collectedRelationships,
        encounteredTypes: context.jobState.encounteredTypes,
      }).toMatchSnapshot();

      const apiKeyEntities = context.jobState.collectedEntities.filter((e) =>
        e._class.includes('AccessKey'),
      );

      expect(apiKeyEntities.length).toBeGreaterThan(0);
      expect(apiKeyEntities).toMatchGraphObjectSchema({
        _class: ['AccessKey'],
        schema: {
          additionalProperties: false,
          properties: {
            _type: { const: 'aquasec_api_key' },
            _rawData: {
              type: 'array',
              items: { type: 'object' },
            },
            id: { type: 'string' },
            name: { type: 'string' },
            description: { type: 'string' },
            whitelisted: { type: 'boolean' },
            scansPerMonth: { type: 'number' },
            accountId: { type: 'number' },
            groupId: { type: 'number' },
            iacToken: { type: 'boolean' },
            ipAddresses: {
              type: 'array',
              items: { type: 'string' },
            },
          },
        },
      });
    }),
  );
});
