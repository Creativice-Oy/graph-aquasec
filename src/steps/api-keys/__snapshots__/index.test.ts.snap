// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchApiKeys should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "aquasec_account:10630",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "additional_credits": 0,
            "associate_tags": false,
            "aws_marketplace_id": "",
            "block_default_group": false,
            "byok_bucket": null,
            "created": "2021-10-09T00:36:24.000Z",
            "csp_enabled": true,
            "csp_metadata": null,
            "current_plan": 57,
            "custom_cost": null,
            "disable_default_emails": false,
            "enforce_oauth": false,
            "global_suppressions_admin_only": false,
            "id": 10630,
            "integrations_admin_only": false,
            "just_in_time_domain": null,
            "just_in_time_domain_secondary": null,
            "login_timeout": null,
            "num_allowed_keys": 1,
            "presuppress_plugins": false,
            "renewal_date": null,
            "restrict_to_ips": null,
            "root_user": "austin.kelleher@jupiterone.dev",
            "source": "aquasec",
            "stripe_id": null,
            "suppressions_admin_only": false,
            "trial_end": "2021-10-23T00:36:24.000Z",
            "users_admin_only": false,
          },
        },
      ],
      "_type": "aquasec_account",
      "additionalCredits": 0,
      "associateTags": false,
      "awsMarketplaceId": "",
      "blockDefaultGroup": false,
      "createdOn": 1633739784000,
      "cspEnabled": true,
      "currentPlan": 57,
      "disableDefaultEmails": false,
      "displayName": "10630",
      "enforceOauth": false,
      "globalSuppressionsAdminOnly": false,
      "id": "10630",
      "integrationsAdminOnly": false,
      "manager": "austin.kelleher@jupiterone.dev",
      "name": "10630",
      "numAllowedKeys": 1,
      "presuppressPlugins": false,
      "rootUser": "austin.kelleher@jupiterone.dev",
      "source": "aquasec",
      "stripeId": undefined,
      "suppressionsAdminOnly": false,
      "trialEnd": 1634949384000,
      "usersAdminOnly": false,
    },
    Object {
      "_class": Array [
        "AccessKey",
      ],
      "_key": "aquasec_api_key:1081",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_key": "mebDoeHcCHm6RcCraAaqgP",
            "account_id": 10630,
            "created": "2021-10-09T12:53:56.000Z",
            "description": null,
            "group_id": null,
            "iac_token": false,
            "id": 1081,
            "ip_addresses": null,
            "scans_per_month": 0,
            "whitelisted": false,
          },
        },
      ],
      "_type": "aquasec_api_key",
      "accountId": 10630,
      "createdOn": 1633784036000,
      "description": undefined,
      "displayName": "mebDoeHcCHm6RcCraAaqgP",
      "groupId": undefined,
      "iacToken": false,
      "id": "1081",
      "ipAddresses": undefined,
      "name": "mebDoeHcCHm6RcCraAaqgP",
      "scansPerMonth": 0,
      "whitelisted": false,
    },
    Object {
      "_class": Array [
        "AccessKey",
      ],
      "_key": "aquasec_api_key:1083",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_key": "3FqDYAG325qYx6zN8GZ4Bx",
            "account_id": 10630,
            "created": "2021-10-09T22:00:15.000Z",
            "description": "J1 Limited Scope Key",
            "group_id": 12028,
            "iac_token": false,
            "id": 1083,
            "ip_addresses": null,
            "scans_per_month": 0,
            "whitelisted": false,
          },
        },
      ],
      "_type": "aquasec_api_key",
      "accountId": 10630,
      "createdOn": 1633816815000,
      "description": "J1 Limited Scope Key",
      "displayName": "3FqDYAG325qYx6zN8GZ4Bx",
      "groupId": 12028,
      "iacToken": false,
      "id": "1083",
      "ipAddresses": undefined,
      "name": "3FqDYAG325qYx6zN8GZ4Bx",
      "scansPerMonth": 0,
      "whitelisted": false,
    },
    Object {
      "_class": Array [
        "AccessKey",
      ],
      "_key": "aquasec_api_key:1085",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_key": "D2yEcjM3mOMz5HjmSJzvsf",
            "account_id": 10630,
            "created": "2021-10-09T22:10:14.000Z",
            "description": "JupiterOne API Key",
            "group_id": 12031,
            "iac_token": false,
            "id": 1085,
            "ip_addresses": null,
            "scans_per_month": 0,
            "whitelisted": false,
          },
        },
      ],
      "_type": "aquasec_api_key",
      "accountId": 10630,
      "createdOn": 1633817414000,
      "description": "JupiterOne API Key",
      "displayName": "D2yEcjM3mOMz5HjmSJzvsf",
      "groupId": 12031,
      "iacToken": false,
      "id": "1085",
      "ipAddresses": undefined,
      "name": "D2yEcjM3mOMz5HjmSJzvsf",
      "scansPerMonth": 0,
      "whitelisted": false,
    },
    Object {
      "_class": Array [
        "AccessKey",
      ],
      "_key": "aquasec_api_key:1086",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "access_key": "kRhlEeBBqDAUcEmQdMkAiO",
            "account_id": 10630,
            "created": "2021-10-09T23:48:15.000Z",
            "description": "With IPs",
            "group_id": null,
            "iac_token": false,
            "id": 1086,
            "ip_addresses": Array [
              "127.0.0.1",
              "127.0.0.2",
            ],
            "scans_per_month": 0,
            "whitelisted": false,
          },
        },
      ],
      "_type": "aquasec_api_key",
      "accountId": 10630,
      "createdOn": 1633823295000,
      "description": "With IPs",
      "displayName": "kRhlEeBBqDAUcEmQdMkAiO",
      "groupId": undefined,
      "iacToken": false,
      "id": "1086",
      "ipAddresses": Array [
        "127.0.0.1",
        "127.0.0.2",
      ],
      "name": "kRhlEeBBqDAUcEmQdMkAiO",
      "scansPerMonth": 0,
      "whitelisted": false,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_account:10630",
      "_key": "aquasec_account:10630|has|aquasec_api_key:1081",
      "_toEntityKey": "aquasec_api_key:1081",
      "_type": "aquasec_account_has_api_key",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_account:10630",
      "_key": "aquasec_account:10630|has|aquasec_api_key:1083",
      "_toEntityKey": "aquasec_api_key:1083",
      "_type": "aquasec_account_has_api_key",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_account:10630",
      "_key": "aquasec_account:10630|has|aquasec_api_key:1085",
      "_toEntityKey": "aquasec_api_key:1085",
      "_type": "aquasec_account_has_api_key",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "aquasec_account:10630",
      "_key": "aquasec_account:10630|has|aquasec_api_key:1086",
      "_toEntityKey": "aquasec_api_key:1086",
      "_type": "aquasec_account_has_api_key",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "aquasec_account",
    "aquasec_api_key",
    "aquasec_account_has_api_key",
  ],
  "numCollectedEntities": 5,
  "numCollectedRelationships": 4,
}
`;
